### PR TITLE
Fixed #22804 -- Warn on unsafe value of 'sep=' in Signer

### PR DESCRIPTION
**Pull request by jaap3**

### CODE REVIEW

File: django/core/signing.py [hunk 1]
```diff
@@ -151,6 +155,14 @@ class Signer(object):
     def __init__(self, key=None, sep=':', salt=None):
         # Use of native strings in all versions of Python
         self.sep = force_str(sep)
+        sep_is_invalid = sep != ":" and (
+            not self.sep or
+            self.sep[0] in _SEP_UNSAFE or
```
Comment (timgraham) [inline django/core/signing.py]: why checking only the first and last values of `self.sep`? how about `any(sep in _SEP_UNSAFE for sep in self.sep)`?

File: django/core/signing.py [hunk 2]
```diff
@@ -38,15 +38,19 @@
...
 import json
+import string
 import time
+import warnings
 import zlib
 
 from django.conf import settings
 from django.utils import baseconv
 from django.utils.crypto import constant_time_compare, salted_hmac
+from django.utils.deprecation import RemovedInNextVersionWarning
```
Comment (timgraham) [inline django/core/signing.py]: should use `RemovedInDjango110Warning`

**Comment by timgraham** [comment]
See [Deprecating a feature](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature) for documentation requirements.

**Comment by jaap3** [comment]
I've updated the pull request. Changed the validity test to a regex based one. The separator should not be indistinguishable from the signature, so as long as it contains a character outside the range of urlsafe characters it should be fine.

Updated the docs. There already is a note about the separator on https://docs.djangoproject.com/en/1.8/topics/signing/#django.core.signing.Signer so I didn't expand on that.

File: django/core/signing.py [hunk 3]
```diff
@@ -150,11 +155,23 @@ class Signer(object):
...
 
+    @property
+    def sep(self):
+        return self._sep
+
+    @sep.setter
+    def sep(self, value):
+        sep = force_str(value)
+        if _SEP_UNSAFE.match(sep):
+            warnings.warn('Unsafe Signer separator: %r (cannot be empty or consist of only A-z0-9-_=)' % sep,
```
Comment (timgraham) [inline django/core/signing.py]: are you sure about this point of "consist of only A-z0-9-_=", as opposed to those characters may not appear in the separator? (not too familiar with this restriction myself)
I don't know that the complexity of a `setter` is necessary. I'd just modify the test to initialize a `Signer` each time.

File: tests/signing/tests.py [hunk 4]
```diff
@@ -112,6 +113,26 @@ def test_works_with_non_ascii_keys(self):
...
+        """should be able to change the separator"""
+        separators = ['/', '*sep*', ',']
+        signer = signing.Signer('predictable-secret')
+        for sep in separators:
+            signer.sep = sep
+            self.assertEqual('foo%ssH9B01cZcJ9FoT_jEVkRkNULrl8' % sep, signer.sign('foo'))
+
+    def test_invalid_sep(self):
+        """should warn on invalid separator"""
+        separators = ["", "-", "abc",]
```
Comment (timgraham) [inline tests/signing/tests.py]: please use single quotes for consistency

**Comment by timgraham** [comment]
doc edits: http://dpaste.com/20NM7T4 (please wrap at 79 chars)

File: django/core/signing.py (hunk 3 above)
Comment (jaap3) [inline django/core/signing.py]: sep is used in as an argument to `rsplit` in `unsign` like so:

    value, sig = signed_value.rsplit(self.sep, 1)

So as long as the separator contains at least one character that will never appear in the signature string it should be safe.

I can remove the `setter`. My reasoning behind it is that `Signer().sep` is (undocumented, but) public, so it should be checked at all times. Same could be said for key and salt. I do agree that it bloats the code a fair bit. So I guess it's your call.

Comment (timgraham) [inline django/core/signing.py]: I agree if we wanted to add the `setter`, we should have it for `salt` as well and I don't see much benefit to the extra complexity.

**Comment by jaap3** [comment]
Updated the PR

**Comment by timgraham** [comment]
merged in 0d71349773f8d2e8160d76b5c91a1896ece772db, thanks!

**Pull request closed without merging**